### PR TITLE
feat: added z_algorithm in strings

### PR DESCRIPTION
**Pull request by RitikaGupta8734**
#### Description of Change
Added Z-Algorithm for pattern matching in strings.

#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: 
Kindly review my PR and let me know any corrections.

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1581"></a>

### CODE REVIEW

**Comment by RitikaGupta8734** [comment]
@Panquesito7  Kindly look into my PR.

**Review by ayaankhan98 — CHANGES_REQUESTED** [review]
please look at the [contribution guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md)

**Comment by RitikaGupta8734** [comment]
Hi @ayaankhan98 Can you be more specific where I made mistake.

File: strings/z_function.cpp [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+ * @brief The Z function for finding occurences of a pattern 
+ * within a piece of text with time and space complexity O(n + m)
+ * @details 
+ * DESCRIPTION
+ * 1. The Z-function for a string is an array of length n where the 
+ * i-th element is equal to the greatest number of characters starting 
+ * from the position i that coincide with the first characters of s.
+ * 2. Eg string : ababb then z[2]=2 as s[2]=s[0] and s[3]=s[1] and s[4]!=s[2]
+ * LINK
+ * https://cp-algorithms.com/string/z-function.html
```
Comment (ayaankhan98) [inline strings/z_function.cpp:12]: use github markdown syntax for links

File: strings/z_function.cpp [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    std::string pattern1 = "abc";
+    
+    std::vector<int> matching_indexes1=find_pat_in_text(pattern1,text1);
+    assert((matching_indexes1 == std::vector<int>{10,14,18,30,33}));
+    
+    std::string text2 = "greengrass";
+    std::string pattern2 = "abc";
+    
+    std::vector<int> matching_indexes2=find_pat_in_text(pattern2,text2);
+    assert((matching_indexes2 == std::vector<int>{}));
```
Comment (ayaankhan98) [inline strings/z_function.cpp:72]: create a separate function for tests and try to cover all test cases (corner cases).

File: strings/z_function.cpp (hunk 1 above)
Comment (RitikaGupta8734) [reply strings/z_function.cpp]: Kindly have a look at the updated file.

File: strings/z_function.cpp (hunk 2 above)
Comment (RitikaGupta8734) [reply strings/z_function.cpp]: Hi I have made the required changes. Kindly have a look.

File: strings/z_function.cpp [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    std::vector<int> z = Z_function(pattern+'#'+text);
+    std::vector<int> matching_indexes;
+
+    for (int i = 0; i < text_length; i++) {
+        if(z[i+pattern_length+1]==pattern_length) matching_indexes.push_back(i);   
+    }
+    return matching_indexes;
+}  
+
+void testing(){
```
Comment (Panquesito7) [inline strings/z_function.cpp:57]:
```suggestion
/**
 * @brief Self-test implementations
 * @returns void
 */
static void test(){
```

File: strings/z_function.cpp [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    
+    // corner case
+    std::string text2 = "greengrass";
+    std::string pattern2 = "abc";
+    
+    std::vector<int> matching_indexes2=find_pat_in_text(pattern2,text2);
+    assert((matching_indexes2 == std::vector<int>{}));
+}
+
+/** Main function */
```
Comment (Panquesito7) [inline strings/z_function.cpp:74]:
```suggestion
/**
 * @brief Main function
 * @returns 0 on exit
 */
```

File: strings/z_function.cpp [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+int main() {
+    
+    testing();
```
Comment (Panquesito7) [inline strings/z_function.cpp:77]:
```suggestion
    test();  // run self-test implementations
```

File: strings/z_function.cpp [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    for (int i = 1,l = 0,r = 0; i < pattern_length; i++) {
+        if(i<=r)z[i]=std::min(r-i+1,z[i-l]);
+        while(i+z[i]<pattern_length && pattern[z[i]]==pattern[i+z[i]]) z[i]++;
+        if(i+z[i]-1>r)r=i+z[i]-1;
+    }
+    return z;
+}
+
+/**
+ * Using Z_function to find a pattern in a text
```
Comment (Panquesito7) [inline strings/z_function.cpp:41]:
```suggestion
 * @brief Using Z_function to find a pattern in a text
```

File: strings/z_function.cpp [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+#ifdef _MSC_VER
+#include <string>  // use this for MS Visual C++
+#else
+#include <cstring>
+#endif
+#include <vector>
+# include <cassert>
+
+/**
+ * Generate the Z-function for the inputted string.
```
Comment (Panquesito7) [inline strings/z_function.cpp:24]:
```suggestion
 * @brief Generate the Z-function for the inputted string.
```

File: strings/z_function.cpp [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+ * @details 
+ * DESCRIPTION
+ * 1. The Z-function for a string is an array of length n where the 
+ * i-th element is equal to the greatest number of characters starting 
+ * from the position i that coincide with the first characters of s.
+ * 2. Eg string : ababb then z[2]=2 as s[2]=s[0] and s[3]=s[1] and s[4]!=s[2]
+ * @author [Ritika Gupta](https://github.com/RitikaGupta8734)
+ */
+
+#include <iostream>
```
Comment (Panquesito7) [inline strings/z_function.cpp:14]:
```suggestion
#include <iostream>   /// for IO operations
```

File: strings/z_function.cpp [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+#ifdef _MSC_VER
+#include <string>  // use this for MS Visual C++
```
Comment (Panquesito7) [inline strings/z_function.cpp:16]:
```suggestion
#include <string>  /// for string (use this for MS Visual C++)
```

File: strings/z_function.cpp [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+#else
+#include <cstring>
```
Comment (Panquesito7) [inline strings/z_function.cpp:18]:
```suggestion
#include <cstring>   /// for string
```

File: strings/z_function.cpp [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+#endif
+#include <vector>
+# include <cassert>
```
Comment (Panquesito7) [inline strings/z_function.cpp:21]:
```suggestion
#include <vector>     /// for std::vector
#include <cassert>   /// for assert
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Great work! 😄👍

**Comment by RitikaGupta8734** [comment]
Hi @Panquesito7 , I have made the suggested changes.

**Comment by RitikaGupta8734** [comment]
> Great work!

Thank you :)

File: strings/z_function.cpp [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+}
+
+/**
+ * @brief Using Z_function to find a pattern in a text
+ * \param[in] pattern string pattern to search
+ * \param[in] text text in which to search
+ * \returns a vector of starting indexes where pattern is found in the text 
+ */
+std::vector<int> find_pat_in_text(const std::string &pattern, const std::string &text) {
+    int text_length = text.size(), pattern_length = pattern.size();
```
Comment (Panquesito7) [inline strings/z_function.cpp:47]: Consider using `uint64_t` for non-negative values (or their appropriate size: `uint32_t`, `uint16_t`, `uint8_t`) or `int64_t` for negative values. Check other parts of the code ([reference](https://en.cppreference.com/w/cpp/types/integer)). 🙂

File: strings/z_function.cpp [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+    }
+    return matching_indexes;
+}  
+
+/**
+ * @brief Self-test implementations
+ * @returns void
+ */
+static void test(){
+
```
Comment (Panquesito7) [inline strings/z_function.cpp:62]:
```suggestion
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Please enable [GitHub Actions](https://docs.github.com/en/free-pro-team@latest/actions/learn-github-actions) in your repository of this fork in this link: https://github.com/RitikaGupta8734/C-Plus-Plus/actions

File: strings/z_function.cpp (hunk 12 above)
Comment (RitikaGupta8734) [reply strings/z_function.cpp]: Made the required changes. Kindly review :)

File: strings/z_function.cpp [hunk 14]
```diff
@@ -0,0 +1,95 @@
+/**
+ * @file
+ * @brief The [Z function](https://cp-algorithms.com/string/z-function.html) for
+ * finding occurences of a pattern within a piece of text with time and space
+ * complexity O(n + m)
+ * @details
+ * DESCRIPTION
```
Comment (Panquesito7) [inline strings/z_function.cpp:7]:
```suggestion
```

File: strings/z_function.cpp [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+ * 1. The Z-function for a string is an array of length n where the
+ * i-th element is equal to the greatest number of characters starting
+ * from the position i that coincide with the first characters of s.
+ * 2. Eg string : ababb then z[2]=2 as s[2]=s[0] and s[3]=s[1] and s[4]!=s[2]
```
Comment (Panquesito7) [inline strings/z_function.cpp:11]:
```suggestion
 * 2. E.g.: string: ababb then z[2]=2 as s[2]=s[0] and s[3]=s[1] and s[4]!=s[2]
```

File: strings/z_function.cpp [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+
+    std::vector<uint64_t> matching_indexes1 = find_pat_in_text(pattern1, text1);
+    assert((matching_indexes1 == std::vector<uint64_t>{10, 14, 18, 30, 33}));
+
+    // corner case
+    std::string text2 = "greengrass";
+    std::string pattern2 = "abc";
+
+    std::vector<uint64_t> matching_indexes2 = find_pat_in_text(pattern2, text2);
+    assert((matching_indexes2 == std::vector<uint64_t>{}));
```
Comment (Panquesito7) [inline strings/z_function.cpp:85]: Please add documentation of what this code is doing here. 🙂

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Great work! Almost there. 😄

File: strings/z_function.cpp (hunk 16 above)
Comment (RitikaGupta8734) [reply strings/z_function.cpp]: Added comments for the same.

File: strings/z_function.cpp [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+/**
+ * @brief Self-test implementations
+ * @returns void
+ */
+static void test() {
+    // usual case
+    std::string text1 = "alskfjaldsabc1abc1abcbksbcdnsdabcabc";
+    std::string pattern1 = "abc";
+
+    //  matching_indexes1 gets the indexes where pattern1 exists in text1
```
Comment (Panquesito7) [inline strings/z_function.cpp:76]:
```suggestion
    // matching_indexes1 gets the indexes where pattern1 exists in text1
```

File: strings/z_function.cpp [hunk 18]
```diff
@@ -0,0 +1,96 @@
...
+    std::vector<uint64_t> matching_indexes1 = find_pat_in_text(pattern1, text1);
+    assert((matching_indexes1 == std::vector<uint64_t>{10, 14, 18, 30, 33}));
+
+    // corner case
```
Comment (Panquesito7) [inline strings/z_function.cpp:80]:
```suggestion
    // corner case
```

File: strings/z_function.cpp [hunk 19]
```diff
@@ -0,0 +1,96 @@
...
+    std::string text2 = "greengrass";
+    std::string pattern2 = "abc";
+
+    //  matching_indexes2 gets the indexes where pattern2 exists in text2
```
Comment (Panquesito7) [inline strings/z_function.cpp:84]:
```suggestion
    // matching_indexes2 gets the indexes where pattern2 exists in text2
```

**Review by Panquesito7 — COMMENTED** [review]
Almost there! It's just to accept these suggestions, and you're done! 😄🎉

File: strings/z_function.cpp (hunk 18 above)
Comment (RitikaGupta8734) [reply strings/z_function.cpp:80]: Unable to accept this one as it shows no changes were made to the code.

Comment (Panquesito7) [reply strings/z_function.cpp:80]: Sorry 😅 probably got confused or something.

**Review by Panquesito7 — APPROVED** [review]
Awesome work, @RitikaGupta8734! It's been awesome for your first contribution here! We hope you keep contributing! 😄👍🎉

**Comment by RitikaGupta8734** [comment]
> Awesome work, @RitikaGupta8734! It's been awesome for your first contribution here! We hope you keep contributing!

Thanks a lot @Panquesito7 :)

**Review by mishraabhinn — APPROVED** [review]

File: strings/z_function.cpp (hunk 18 above)
Comment (RitikaGupta8734) [reply strings/z_function.cpp:80]: Ya np :)

**Comment by RitikaGupta8734** [comment]
Hi @ayaankhan98 kindly review the PR again and let me know the feedback if any.

**Comment by RitikaGupta8734** [comment]
Hi @Panquesito7 will my PR be merged?

**Review by ayaankhan98 — APPROVED** [review]
LGTM :+1:

**Comment by RitikaGupta8734** [comment]
Thank You all ! :)
